src/common/snippets/src/pass/lowered/loop_init.cpp

```diff
@@ -0,0 +1,234 @@
+
+    auto is_buffer_input = [&linear_ir](const LoweredExprPtr& expr) {
+        const auto parent_expr = linear_ir.get_expr_by_output(expr->get_inputs().front()).first;
+        return ov::is_type<op::Buffer>(parent_expr->get_node());
+    };
+    auto is_buffer_output = [&linear_ir](const LoweredExprPtr& expr) {
+        const auto child_exprs_inputs = linear_ir.get_exprs_by_input(expr->get_outputs().front());
+        return ov::is_type<op::Buffer>((*child_exprs_inputs.begin()).first->get_node());
+    };
+    auto there_is_buffer = std::any_of(loop_in_exprs.begin(), loop_in_exprs.end(), is_buffer_input) ||
+                           std::any_of(loop_out_exprs.begin(), loop_out_exprs.end(), is_buffer_output);
+
+    const auto& loop_end = std::make_shared<op::LoopEnd>(
+            managed_outputs, work_amount, work_amount_increment, ptr_increments, finalization_offsets);
+    loop_end->set_work_with_buffer(there_is_buffer);
+    loop_end->has_outer_loop = has_outer_loop;
+
+    std::vector<TensorDescriptorPtr> loop_end_inputs;
+    for (const auto& expr : loop_in_exprs)
+        loop_end_inputs.push_back(expr->get_inputs().front());
+    for (const auto& expr : loop_out_exprs)
+        loop_end_inputs.push_back(expr->get_outputs().front());
+    loop_end_inputs.push_back(linear_ir.get_expr_by_node(loop_begin)->get_outputs().front());
+
+    const auto& loop_end_expr = std::make_shared<LoweredExpr>(loop_end, loop_end_inputs);
+    linear_ir.insert(loop_end_pos, loop_end_expr);
+    return true;
+}
+
+bool LoopInit::run(LoweredExprIR& linear_ir) {
+    OV_ITT_SCOPED_TASK(ngraph::pass::itt::domains::SnippetsTransform, "Snippets::LoopInit")
+    if (linear_ir.empty())
+        return false;
+
+    const auto& loop_manager = linear_ir.get_loop_manager();
+
+    std::set<size_t> inserted_loops;
+    std::set<size_t> loop_identifies = loop_manager->get_identifies();
+    for (auto expr_it = linear_ir.begin(); expr_it != linear_ir.end(); expr_it++) {
+        const auto expr = *expr_it;
+        const auto& node = expr->get_node();
+        if (ov::is_type<op::LoopBase>(node) ||
+            ov::is_type<op::Buffer>(node) ||     // Need to cover Buffer
+            ov::is_type<opset1::Parameter>(node) ||
+            ov::is_type<opset1::Result>(node))
+            continue;
+
+        // Outer Loop ----> Inner Loop
+        const auto expr_loops = expr->get_loop_identifies();
+        const auto loop_depth = expr_loops.size();
+        for (size_t i = 0; i < loop_depth; ++i) {
+            const auto loop_id = expr_loops[i];
+            if (loop_id == LoweredExpr::LOOP_NULL_ID)
+                continue;
+            bool need_to_insert = loop_identifies.find(loop_id) != loop_identifies.end();
+            if (need_to_insert) {
+                const auto loop_info = loop_manager->get(loop_id);
+                const bool has_outer_loop = i > 0 && inserted_loops.find(expr_loops[i - 1]) != inserted_loops.end();
```

Comment:
It looks like here we rely on some properties of loop numbering (e.g. [outer id] == [inner_id - 1]), do we? But we don't control that these properties are not violated somewhere in pipeline, so this might get us into trouble.
Can't we just check that the current loop id  is the first one in an expression loop_ids?